azurerm/internal/services/sentinel/sentinel_data_connector_threat_intelligence.go

```diff
@@ -0,0 +1,168 @@
+	client := meta.(*clients.Client).Sentinel.DataConnectorsClient
+	ctx, cancel := timeouts.ForCreateUpdate(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	workspaceId, err := loganalyticsParse.LogAnalyticsWorkspaceID(d.Get("log_analytics_workspace_id").(string))
+	if err != nil {
+		return err
+	}
+	name := d.Get("name").(string)
+	id := parse.NewDataConnectorID(workspaceId.SubscriptionId, workspaceId.ResourceGroup, workspaceId.WorkspaceName, name)
+
+	if d.IsNewResource() {
+		resp, err := client.Get(ctx, id.ResourceGroup, operationalInsightsResourceProvider, id.WorkspaceName, name)
+		if err != nil {
+			if !utils.ResponseWasNotFound(resp.Response) {
+				return fmt.Errorf("checking for existing Sentinel Data Connector Threat Intelligence %q: %+v", id, err)
+			}
+		}
+
+		id := dataConnectorID(resp.Value)
+		if id != nil && *id != "" {
+			return tf.ImportAsExistsError("azurerm_sentinel_data_connector_threat_intelligence", *id)
+		}
+	}
+
+	tenantId := d.Get("tenant_id").(string)
+	if tenantId == "" {
+		tenantId = meta.(*clients.Client).Account.TenantId
+	}
+
+	param := securityinsight.TIDataConnector{
+		Name: &name,
+		TIDataConnectorProperties: &securityinsight.TIDataConnectorProperties{
+			TenantID: &tenantId,
+			DataTypes: &securityinsight.TIDataConnectorDataTypes{
+				Indicators: &securityinsight.TIDataConnectorDataTypesIndicators{
+					State: securityinsight.Enabled,
+				},
+			},
+		},
+		Kind: securityinsight.KindThreatIntelligence,
+	}
+
+	_, err = client.CreateOrUpdate(ctx, id.ResourceGroup, operationalInsightsResourceProvider, id.WorkspaceName, id.Name, param)
+	if err != nil {
+		return fmt.Errorf("creating Sentinel Data Connector Threat Intelligence %q: %+v", id, err)
+	}
+
+	d.SetId(id.ID())
+
+	return resourceSentinelDataConnectorThreatIntelligenceRead(d, meta)
+}
+
+func resourceSentinelDataConnectorThreatIntelligenceRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*clients.Client).Sentinel.DataConnectorsClient
+	ctx, cancel := timeouts.ForRead(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	id, err := parse.DataConnectorID(d.Id())
+	if err != nil {
+		return err
+	}
+	workspaceId := loganalyticsParse.NewLogAnalyticsWorkspaceID(id.SubscriptionId, id.ResourceGroup, id.WorkspaceName)
+
+	resp, err := client.Get(ctx, id.ResourceGroup, operationalInsightsResourceProvider, id.WorkspaceName, id.Name)
+	if err != nil {
+		if utils.ResponseWasNotFound(resp.Response) {
+			log.Printf("[DEBUG] Sentinel Data Connector Threat Intelligence %q was not found - removing from state!", id)
+			d.SetId("")
+			return nil
+		}
+
+		return fmt.Errorf("retrieving Sentinel Data Connector Threat Intelligence %q: %+v", id, err)
+	}
+
+	if err := assertDataConnectorKind(resp.Value, securityinsight.DataConnectorKindThreatIntelligence); err != nil {
+		return fmt.Errorf("asserting Sentinel Data Connector Threat Intelligence of %q: %+v", id, err)
+	}
+	dc := resp.Value.(securityinsight.TIDataConnector)
```

Comment:
tbh we could likely remove the `assert` call here and make this:

```suggestion
	dc, ok := resp.Value.(securityinsight.TIDataConnector)
	if !ok {
		return fmt.Errorf("%s was not a Threat Intelligence Connector", *id)
	}
```

Reply:
The assert function can be used both in Read function and the `importer`. For the sake of not to repeat the type assertion among all the DC files, I'd rather centralize the logic in one place.